README.md
# Functionality Adding Project

This project is constructed to add a web functionality using Flask for backend and Vue.js for frontend using data offered by Torre's API. The app will consist of 3 new functionalities that are implemented in the backend, and that can be used from the frontend found at the https://github.com/rafaelfrieri1/FunctionalityAddProjectFront/tree/master repository.

- The first functionality will be "Trending Skills" which will consist of a small dashboard showing which are the most seeked skills by job opportunities at Torre.co. Many people can't find a job easily and would like to learn what jobs are expecting for them to know, then a good starting point could be looking at the statistics (They don't lie!).

- The second functionality will be the "Search for Recruiters" functionality. Among recruiters, one of the most time taking tasks is searching for job talent. That's why, with this functionality recruiters can introduce a job ID and receive a list of possible candidates that could fit for the job (based on their skills). This list of candidates is extracted from a random sample (sized 10000) from the users database at Torre.co.

- The third functionality will be "Unpopular Audiences". This third functionality permits the visualization of the most unpopular areas and/or skills at torre.co. As it is intended to attract more people daily to the platform, it is needed to know which are the weakest audiences so future platform enhances are oriented on covering a bigger market.

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## Technology
For the backend code the Flask from Python framework was used.

## License
[MIT](https://choosealicense.com/licenses/mit/)

CHANGELOG.md
# Change Log

## [1.0.5] 2020-12-09
- Minor changes on code
- Wrote a more complete README file

## [1.0.4] 2020-12-09
- Unpopular Audiences implemented
- Tests performed on the new functionality
- Minor changes performed

## [1.0.3] 2020-12-08
- Recruiter's Search implemented
- Algorithm analyzes 10000 of a random sample in the people's search API
- Tests performed, results are good most of the times. (Can last 30 seconds)

## [1.0.2] 2020-12-08
- Trending skills functionality set
- Libraries such as numpy included for a faster algorith
- Tests done with Torre.co API

## [1.0.1] 2020-12-08
### Test
- Test blueprint set


## [1.0.0] 2020-12-08
### Flask project template stablished.

application/Test/Test.py
from flask import Blueprint
from flask import current_app as app

test_bp = Blueprint('test_bp', __name__)

@test_bp.route('/', methods = ['GET'])

def test():
    return "Hello World!"

application/popular_skills/popular_skills.py
import requests
import numpy as np

from flask import Blueprint, request, make_response, jsonify
from flask import current_app as app

popular_skills_bp = Blueprint('popular_skills_bp', __name__)

@popular_skills_bp.route('/popular-skills/', methods = ['GET'])
def popular_skills():
    if request.method != 'GET' :
        return make_response('Wrong query performed.', 400)

    paramsReq = {'offset':'0', 'size':'0', 'aggregate':'true'}
    resp = requests.post('https://search.torre.co/opportunities/_search', params = paramsReq)

    if resp.status_code == 400:
        return make_response('Information API of job search is not responding.', 400)

    n = int(request.args.get('size'))
    skills = resp.json()['aggregators']['skill']
    amount = np.array([])
    name = np.array([])

    for skill in skills:

        if len(amount) < n:
            amount = np.append(amount, skill['total'])
            name = np.append(name, skill['value'])
        
        if (min(amount) < skill['total']) and (len(amount) == n):
            minIndexRes = np.where(amount == min(amount))
            minIndex = minIndexRes[0]
            amount[minIndex[0]] = skill['total']
            name[minIndex[0]] = skill['value']

    sortIndexes = np.argsort(-amount, kind = 'mergesort')

    amountSent = np.array([])
    nameSent = np.array([])
    percentageSent = np.array([])
    totalJobOpportunities = resp.json()['total']

    for index in sortIndexes:
        amountSent = np.append(amountSent, amount[index])
        nameSent = np.append(nameSent, name[index])
        percentageIndex = (amount[index]/totalJobOpportunities)*100
        percentageSent = np.append(percentageSent, round(percentageIndex, 2))


    pop_skills = {'total': amountSent.tolist(), 'value': nameSent.tolist(), 'percentage_jobs': percentageSent.tolist()}
    headers = {'Content-Type': 'applicaton/json', 'Access-Control-Allow-Origin':'https://torre-application-front.herokuapp.com'}

    return make_response(jsonify(pop_skills), 200, headers)

application/target_audiences_platform_enhance/target_audiences_platform_enhance.py
import requests
import numpy as np

from flask import Blueprint, request, make_response, jsonify
from flask import current_app as app

target_audiences_platform_enhance_bp = Blueprint('target_audiences_platform_enhance_bp', __name__)

@target_audiences_platform_enhance_bp.route('/target-audiences-platform-enhance/', methods = ['GET'])
def target_audiences_platform_enhance():
    if request.method != 'GET' :
        return make_response('Wrong query performed.', 400)

    paramsReq = {'offset':'0', 'size':'0', 'aggregate':'true'}
    resp = requests.post(' https://search.torre.co/people/_search', params = paramsReq)

    if resp.status_code == 400:
        return make_response('Information API of job search is not responding.', 400)

    n = int(request.args.get('size'))
    skills = resp.json()['aggregators']['skill']
    amount = np.array([])
    name = np.array([])

    for skill in skills:

        if len(amount) < n:
            amount = np.append(amount, skill['total'])
            name = np.append(name, skill['value'])
        
        if (max(amount) > skill['total']) and (len(amount) == n):
            maxIndexRes = np.where(amount == max(amount))
            maxIndex = maxIndexRes[0]
            amount[maxIndex[0]] = skill['total']
            name[maxIndex[0]] = skill['value']

    sortIndexes = np.argsort(amount, kind = 'mergesort')

    amountSent = np.array([])
    nameSent = np.array([])
    percentageSent = np.array([])
    totalAmountOfPeople = resp.json()['total']

    for index in sortIndexes:
        amountSent = np.append(amountSent, amount[index])
        nameSent = np.append(nameSent, name[index])
        percentageIndex = (amount[index]/totalAmountOfPeople)*100
        percentageSent = np.append(percentageSent, round(percentageIndex, 3))


    unpopular_audience = {'total': amountSent.tolist(), 'value': nameSent.tolist(), 'percentage_jobs': percentageSent.tolist()}
    headers = {'Content-Type': 'applicaton/json', 'Access-Control-Allow-Origin':'https://torre-application-front.herokuapp.com'}

    return make_response(jsonify(unpopular_audience), 200, headers)

application/strength_based_search/strength_based_search.py
import requests
import numpy as np

from flask import Blueprint, request, make_response, jsonify
from flask import current_app as app
from random import seed, randint

strength_based_search_bp = Blueprint('strength_based_search_bp', __name__)

def employeeSearch(employees, topEmployees, topUsernames, topScores, imageURLs, requiredStrengths, n):
    for employee in employees:
        if len(topEmployees) < n :
            topEmployees = np.append(topEmployees, employee["name"])
            topUsernames = np.append(topUsernames, employee["username"])
            topScores = np.append(topScores, score(employee, requiredStrengths))
            imageURLs = np.append(imageURLs, employee["picture"])
        
        if (len(topEmployees)) == n and score(employee, requiredStrengths)>min(topScores):
            minIndexRes = np.where(topScores == min(topScores))
            minIndex = minIndexRes[0]
            topEmployees[minIndex[0]] = employee["name"]
            topUsernames[minIndex[0]] = employee["username"]
            topScores[minIndex[0]] = score(employee, requiredStrengths)
            imageURLs[minIndex[0]] = employee["picture"]

    return [topEmployees, topUsernames, topScores, imageURLs]


def score(employee, requiredStrengths):
    empSkills = employee["skills"]
    scoreResult = 0
    for skill in empSkills:
        for strength in requiredStrengths:
            if skill["name"] == strength["name"] :
                scoreResult = scoreResult + skill["weight"] + 1
                break
    return scoreResult

@strength_based_search_bp.route('/strength-based-search/', methods = ['GET'])
def strength_based_search():
    if request.method != 'GET' :
        return make_response('Wrong query performed.', 400)

    jobID = request.args.get('job-id')
    n = int(request.args.get('size'))
    respOp = requests.get(f'https://torre.co/api/opportunities/{jobID}')
    if respOp.status_code != 200:
        return make_response('Information API of job opportunity with id is not responding.', 400)

    requiredStrengths = respOp.json()['strengths']
    first = True
    seed()
    randInitialize = randint(0, 400001)
    offset = randInitialize
    employees = []
    topEmployees = np.array([])
    topUsernames = np.array([])
    topScores = np.array([])
    imageURLs = np.array([])

    while ((len(employees) == 5000) or (first == True)) and offset != (randInitialize+10000):
        respEmployees = requests.post(f'https://search.torre.co/people/_search/?offset={offset}&size=5000&aggregate=false')
        if respEmployees.status_code == 400:
            return make_response('Information API of people search is not responding.', 400)
        
        employees = respEmployees.json()['results']
        resEmpSearch = employeeSearch(employees, topEmployees, topUsernames, topScores, imageURLs, requiredStrengths, n)
        topEmployees = resEmpSearch[0]
        topUsernames = resEmpSearch[1]
        topScores = resEmpSearch[2]
        imageURLs = resEmpSearch[3]

        if first == True:
            first = False

        offset = offset + 5000

    sortIndexes = np.argsort(-topScores, kind = 'mergesort')

    employeeSent = np.array([])

    for index in sortIndexes:
        employeeObj = {'name': topEmployees[index], 'username': topUsernames[index], 'score': topScores[index], 'picture': imageURLs[index]}
        employeeSent = np.append(employeeSent, employeeObj)

    best_employees = {'employees': employeeSent.tolist(), 'jobName': respOp.json()['serpTags']['title']}
    headers = {'Content-Type': 'applicaton/json', 'Access-Control-Allow-Origin':'https://torre-application-front.herokuapp.com'}

    return make_response(jsonify(best_employees), 200, headers)